src/swarm/neo/request/RequestEventDispatcher.d

```diff
@@ -0,0 +1,777 @@
+    Const!(void)[] payload;
+}
+
+/*******************************************************************************
+
+    A signal which occurred.
+
+*******************************************************************************/
+
+public struct Signalled
+{
+    /// Code of the signal.
+    ubyte code;
+}
+
+/*******************************************************************************
+
+    Struct representing the successful sending of a payload over the connection.
+
+*******************************************************************************/
+
+public struct Sent
+{
+    // Dummy struct with no data, used purely as a "flag" in the smart union to
+    // indicate the event which occurred. (As each fiber is only allowed to
+    // send one thing at a time (see RequestEventDispatcher.register), there is
+    // no need to for this struct to have any fields.
+}
+
+/*******************************************************************************
+
+    Request event dispatcher for use with multi-fiber requests.
+
+    Three types of events are handled:
+        1. "Signals" (ubyte codes) sent between fibers.
+        2. Messages received over the connection.
+        3. Connection readiness to send a payload.
+
+    The types of events are handled somewhat differently, as follows:
+        1. Only one request handler fiber may register to be notified of each
+           type of signal.
+        2. Messages reveived over the connection are expected to begin with a
+           byte representing the message type. Only one request handler fiber
+           may register to be notified of each type of message.
+        3. Each request handler fiber may only register once (at a time) for
+           notification of connection send readiness. Send readiness
+           notifications are handled in order of registration (i.e. fibers will
+           be notified of successive send readiness in the order in which they
+           registered to receive this notification.)
+
+*******************************************************************************/
+
+public struct RequestEventDispatcher
+{
+    import ocean.core.Enforce;
+    import Array = ocean.core.Array : moveToEnd, removeShift, contains;
+    import swarm.neo.util.MessageFiber;
+    import swarm.neo.connection.RequestOnConnBase;
+
+    /// Struct wrapping a waiting fiber along with the event it is awaiting.
+    private struct WaitingFiber
+    {
+        /// Fiber which is waiting.
+        MessageFiber fiber;
+
+        /// Event being awaited.
+        EventRegistration event;
+    }
+
+    /// Fibers currently awaiting events.
+    private WaitingFiber[] waiting_fibers;
+
+    /// Event which has fired, to be returned by nextEvent(). (This value is set
+    /// by notifyWaitingFiber(), just before resuming the waiting fiber.)
+    private EventNotification last_event;
+
+    /// Token passed to fiber suspend/resume calls.
+    private static MessageFiber.Token token;
```

Comment:
This doesn't seem to be initialized anywhere, would this be a problem? (I presume no, as IIRC, we're using `MessageFiber` here solely for its ability to kill the fiber)